bamt/builders/evo_builder.py

```diff
@@ -0,0 +1,195 @@
+
+           Args:
+               data (DataFrame): The data from which to build the structure.
+               classifier (Optional[object]): A classification model for discrete nodes.
+               regressor (Optional[object]): A regression model for continuous nodes.
+
+           Additional optional parameters can be provided as keyword arguments (kwargs) to customize the evolutionary
+           algorithm used to generate the structure. These include parameters to control the size of the population,
+           the probabilities of crossover and mutation, constraints on the structure of the graph, and many others.
+
+           The resulting structure is stored in the `skeleton` attribute of the `EvoStructureBuilder` object.
+        """
+        # Get the list of node names
+        nodes_types = data.columns.to_list()
+
+        # Create the initial population
+        initial = [
+            evo.CustomGraphModel(
+                nodes=kwargs.get(
+                    'init_nodes', [
+                        evo.CustomGraphNode(node_type) for node_type in nodes_types]))]
+
+        # Define the requirements for the evolutionary algorithm
+        requirements = GraphRequirements(
+            max_arity=kwargs.get(
+                'max_arity', self.default_max_arity), max_depth=kwargs.get(
+                'timeout', self.default_max_depth),
+            timeout=timedelta(
+                seconds=kwargs.get(
+                    'timeout', self.default_timeout)),
+            n_jobs=kwargs.get('n_jobs', self.default_n_jobs))
+
+        # Set the parameters for the evolutionary algorithm
+        optimizer_parameters = GPAlgorithmParameters(
+            pop_size=kwargs.get('pop_size', self.default_pop_size),
+            crossover_prob=kwargs.get('crossover_prob', self.default_crossover_prob),
+            mutation_prob=kwargs.get('mutation_prob', self.default_mutation_prob),
+            genetic_scheme_type=GeneticSchemeTypesEnum.steady_state,
+            mutation_types=kwargs.get('custom_mutations', self.default_mutations),
+            crossover_types=kwargs.get('custom_crossovers', self.default_crossovers),
+            selection_types=kwargs.get('selection_type', self.default_selection))
+
+        # Set the adapter for the conversion between the graph and the data
+        # structures used by the optimizer
+        adapter = DirectAdapter(
+            base_graph_class=evo.CustomGraphModel,
+            base_node_class=evo.CustomGraphNode)
+
+        # Set the constraints for the graph
+
+        if kwargs.get('blacklist', None) is not None:
+            self.default_constraints.append(evo.has_no_blacklist_edges)
+        if kwargs.get('whitelist', None) is not None:
+            self.default_constraints.append(evo.has_only_whitelist_edges)
+
+        constraints = kwargs.get(
+            'custom_constraints',
+            self.default_constraints)
```

Comment:
If a person sets custom_constraints, then mandatory constraints must also be added:
             [has_no_self_cycled_nodes,
             has_no_cycle,
             evo.has_no_duplicates]